### DISCUSSION TITLE
Parameter variation with an external optimizer

### DISCUSSION THREAD

**Discussion by warfman98** [the-anylogic-company/AnyLogic-Pypeline#66] [Q&A] [answered]
Hello, 

I'm trying to optimize the parameters of a production planning environment in AnyLogic. To do this, I want to use an external optimizer (genetic algorithm) in Python, where the parameters determined by the algorithm are sent to the AnyLogic parameter variation experiment for evaluation. The output of the AnyLogic simulation run is then the fitness score for the GA. From an earlier discussion it was clear that this value needs to be stored, because the PyCommunicator can't persist between runs.

My question is then, if it would be possible to store this value in the Python script during the parameter variation experiment. In addition, I want to run multiple replications in a simulation run, so the script has to 'wait' until the replications are finished. Is it possible to do such a external optimization using the pypeline connector?

**Comment by t-wolfeadam** [comment, marked as answer]
Hi! Because, to use the Pypeline + the PV experiment. the parameter decisions need to be baked into the script itself in some way, this is seemingly incompatible with an optimizer whose chosen sets of inputs are dynamic over runs.

_Technically_ it can be done by **very** hacky means. As an exercise, I came up with the following: Fix the seed of the GA and only train it on a logged history of the inputs/outputs (starting with training for 0 generations, then 1, etc.). For each unrecorded input, store that in a variable that the pycom can read, give a dummy objective to the GA (so it completes without throwing an error), and have AL update the history with the actual objective, so that the next iteration/generation is correct. Here's a PoC using the ABCA model and the pygad library: [Activity Based Costing Analysis_PythonOptimizer.zip](https://github.com/t-wolfeadam/AnyLogic-Pypeline/files/13620186/Activity.Based.Costing.Analysis_PythonOptimizer.zip) (this doesn't consider replications).

For a more appropriate solution, there's 3 options I can think of, though each has its own limitations/caveats:
* If you have Pro edition, export a simulation experiment to a standalone jar, then write a small java program with command line arguments for the inputs and have it return the result; alternative: have the model override parameters with environment variable values
* If you have a public cloud subscription or a private instance: upload the sim and use the native Python API
* If you can wait ~month, I'm working on overhauling my other library Alpyne, which allows you to use Python to run an exported model for RL purposes (though you could technically set it up without actions).

**Reply by warfman98** [reply]
Thank you so much for your answer and suggestions! Unfortunately, I don't have the pro edition or a public cloud subscription (this project is for my thesis), so I might have to look for a different approach. I will definitely look at your other library as well.
